### ISSUE TITLE
Support multiple amp:access blocks?

### ISSUE THREAD

**Issue by kushal** [ampproject/amphtml#2916]
Possibly crazy, but the two uses I see are:
1) Single sign-on. Especially as users move from device to device, having to reauthenticate with each publishers is a burden, vs. utilizing a user's existing login to any large social network.
2) Authentication-like data from third parties. This is a little more open-ended, but one can imagine ways in which this helps publishers customize articles without needing to plumb data through their own servers.

Of course, this has implications for page weight, but maybe even a small hard limit would mitigate that? On the surface you can also imagine privacy implications, but I'm not sure that's any different than a publisher using its sole amp:access block to reach a third party.

Thoughts?

**Comment by jridgewell** [comment]
Ping @dvoytenko.

**Comment by dvoytenko** [comment]
@kushal We will definitely be looking into ways to extend AMP to a simpler social sign-in protocols. We are closely following Web Credentials API that is currently in active development.

However, a publisher is already allowed to use a 3rd party source for access implementation - all endpoints allow an arbitrary destination as long as it's HTTPS and CORS-enabled endpoints.

**Comment by kushal** [comment]
@dvoytenko I'm actually not familiar with that API, where can I learn more? And the ability to use a third-party source is great - my question is about using *both* a third-party source and a first-party source, or multiple third-party sources, fusing identity from various sources.

**Comment by dvoytenko** [comment]
@kushal Here's the latest version I see. But it's still a bit early - looks like a very actively developed spec. http://w3c.github.io/webappsec-credential-management/

**Comment by kushal** [comment]
Thanks! My quick read also makes it look like federated login is still a ways out.

A related issue is how not all surfaces have access to cookies and the reader ID is scope to the source origin, meaning that even if a third party was providing authentication it could not be reliably carried across sources without the user reauthenticating to the third party each time. I understand the privacy and performance requirements, but it'd be great if there were some escape hatch. Are there any near-term ideas around this?

**Comment by dvoytenko** [comment]
"Reader ID" essentially follows cookie scope: where-ever cookie is available so will be the reader id. But, of course, we don't extend it across browsing boundaries and surfaces. The federated login is expected to help there. We are currently working on the plans to enable it and they are not too far away.

**Comment by kushal** [comment]
I've been chewing this over, but I still think it'd be useful to support multiple amp:access blocks for cases where information can come from both a third-party and the publisher themselves. Even with social login support, it's easy to imagine other scenarios where it's helpful to be able to condition aspects of the display of an article on the results of a RPC to an external service that the user may be authenticated with.

**Comment by dvoytenko** [comment]
@kushal Yes, we are considering this, but it's a little to early to share anything useful. I think your reasoning is good on this, but there are some nuances:

1. For something like this, the authorization response has to be standardized. Otherwise, this would create a mess of trying to reconcile responses with different-meaning fields across multiple endpoints.
2. Authorization endpoints are simple enough - we could run them in parallel and wait for the first one to respond affirmatively. However, login endpoints are harder: it would be a challenge to decide which login to use in which cases.

Do you have some specific ideas about these points?

**Comment by kushal** [comment]
@dvoytenko Thank you so much for considering this! Here are my thoughts.

1. I can definitely see that. My instinctive response would be some sort of namespacing. In the React world, Redux does some default namespacing that people seem to be okay with. Here perhaps there is one canonical amp:access that is not namespaced and all others require some attribute that is used to prefix all resulting variables.

2. Perhaps the same namespace attribute could be used to qualify login links in the body of the document? (For our particular need, we expect our users to already be cookied so we don't need to use the login link. In fact, a totally alternative way of solving this problem for us would be to have an amp:values extension that just loads some attributes from a remote URL using semantics similar to access.)

**Comment by dvoytenko** [comment]
Thanks, @kushal!

> For our particular need, we expect our users to already be cookied so we don't need to use the login link.

Could you identify and describe your case a little more?

**Comment by kushal** [comment]
We're not quite ready to share details, but the general idea is that users who are subscribers to our service should get a different experience (elements shown and hidden, additional analytics) when visiting sites in our network. The user will log in directly on our site before reaching the AMP page and will have a cookie on our domain from that point on, and that cookie will be read by access calls to our domain from AMP pages of sites in our network.

**Comment by dvoytenko** [comment]
Thanks, @kushal! That means that in your case you don't have login multiplicity. Generally, I think we will move in that direction, ignoring login multiplicity initially.

**Comment by adelinamart** [comment]
@dvoytenko are we still considering this? Thank you.

**Comment by kushal** [comment]
FWIW, we'd be more than happy to contribute a custom extension similar to the one for laterpay if that would help it live next to other things.

**Comment by dvoytenko** [comment]
@kushal We are still trying to standardize (or find a standard) for authorization response. This is currently the biggest concern. I don't think namespacing by itself will help, since the results need to be used in access expressions as in 
```
<section amp-access="response.A OR response.B">
```

On the plus side, however, this topic is back in focus for us and we are looking into it very actively again, so I hope to have more updates on this soon.

If you'd like we can bring it on Design Review to discuss if you already have a concrete proposal, of course.

**Comment by kushal** [comment]
@dvoytenko Really glad this is back on your radar and eager to help however I can! I don't totally follow the issue with namespaces. i'm envisioning something like

```
<amp-access namespace="namespaceA">...</amp-access>
<amp-access namespace="namespaceB">...</amp-access>
...
<section amp-access="namespaceA.foo OR namespaceB.bar">..</section>
```

If there was only one amp-access namespace could be omitted and the variables would be unprefixed.

Is there something about AMP that makes this impractical (other than moving away from the singular id="amp-access" markup)?

I think it'll be hard to find a real standard that also deals with the issues around metering? For simple yes/no maybe there's something to leverage from OAuth2?

**Comment by dvoytenko** [comment]
@kushal We've got couple of questions here to address:

1. Expressions. I've also been considering `OR` expressions. They are a bit misleading, but could work. E.g. when one authorization endpoint fails, we can consider all resolutions from that namespace to be `false`.
2. Login. This is a bit harder to arrange from UX point of view. Do you see it as a page markup would offer two login links? E.g. "Login via: <access1> or <access2>"?

**Comment by kushal** [comment]
@dvoytenko Glad we're getting close!

1. Does authorizationFallbackResponse not work here? But no response implying false is also fine for us.

2. Yeah, I was imagining the simplest thing would be to let page markup supply two links using namespaced login links. That gives page owners a good amount of flexibility.

**Comment by dvoytenko** [comment]
Use of both namespaces would work in some instances, however that means that we have to always complete both endpoints and we'll be as slow as the slowest of them. Ideally, we'd like to respond affirmatively as soon as the first endpoint responds affirmatively. 

Another issue I ran into is that some data is more complicated than true/false response. E.g. in metering case, publishers use the response for templates. E.g. `<template>You have ${response.quotaLeft} articles left</template>`. This is a lot harder to manage for multiple namespaces.

OAuth or any other spec do not offer a standard response that I could find.

**Comment by kushal** [comment]
So one disconnect here is that we're imagining that the two auth sources are actually orthogonal, e.g. access source A returns "allowedToHaveDonuts" and access source B "allowedToHaveBeer" and for the most part, parts of the page are conditioned on only one or the other. In that world, rendering should still wait for both responses, possibly with an aggressive timeout and the publisher constructing pieces to have the right default.

Maybe instead of namespaces it would be sufficient to fuse responses and encourage providers to namespace individual variables. If two variables collide deliberately or accidentally, the earliest response could win.

**Comment by ampprojectbot** [comment]
This issue seems to be in  Pending Triage for awhile. @dvoytenko Please triage this to an appropriate milestone.
